### ISSUE TITLE
Type aliases can cause a function to not be equal to its definition

### ISSUE THREAD

**Issue by rosekunkel** [ucsd-progsys/liquidhaskell#1257]
Given these definitions:
```haskell
data Foo = A | B
data Bar = C | D
data Baz = E | F

type Alias = Bar

{-@ reflect f @-}
f :: Foo -> Alias -> Baz
f A _ = E
f B C = F
f _ _ = E
```
the following proof fails:
```haskell
{-@ fProof :: {f B C = F} @-}
fProof :: Proof
fProof =
  f B C === F
  *** QED
```
even though this proof just asserts that `f` is equal to its definition on `B` and `C`. However, if we change the type signature of `f` to
```haskell
f :: Foo -> Bar -> Baz
```
then the proof goes through.

The error given when the proof fails is
```text
Error: Liquid Type Mismatch
  
 19 |   f B C === F
        ^^^^^^^^^^^
  
   Inferred type
     VV : {v : Baz | v == Example.F}
  
   not a subtype of Required type
     VV : {VV : Baz | VV == ?a}
  
   In Context
     ?a : {?a : Baz | ?a == Example.f Example.B Example.C}
```
A complete file demonstrating this is available here: https://gist.github.com/wkunkel/363ccf7a08d6de84390b06838c7b9adf

**Comment by ranjitjhala** [comment]
whoa! that's very odd! Thanks for pointing this out!

**Comment by nikivazou** [comment]
Thanks @wkunkel ! This seems to be a simplified version of https://github.com/ucsd-progsys/liquidhaskell/issues/1225

**Comment by ranjitjhala** [comment]
Digging around, I have simplified @wkunkel 's test case to:

```haskell
{-@ LIQUID "--exactdc" @-}
{-@ LIQUID "--ple"     @-}

module Example where

data Foo = A | B deriving (Eq)
data Bar = C | D deriving (Eq)
data Baz = E | F deriving (Eq)

-- This triggers the bug
-- type Alias = Bar
-- f :: Foo -> Alias -> Baz

-- This renders the program safe
f :: Foo -> Bar -> Baz

{-@ reflect f @-}
f A _ = E
f B C = F
f _ _ = E

{-@ test :: {f B C = F} @-}
test = ()
```

In case *with the alias*  the reflected body is wrong. In the .fq file, we get the *incorrect* definition:

```
define Example.f (ds_d1Vy : Example.Foo,  ds_d1Vz : Example.Bar) = 
  (((Example.f ds_d1Vy ds_d1Vz) = 
    (if (is$Example.A ds_d1Vy) 
     then Example.E 
     else (if true then Example.E else Example.F))))
```

but *without the alias* we get the *correct* definition:

```
define Example.f (ds_d1Vs : Example.Foo,  ds_d1Vt : Example.Bar) = 
  (((Example.f ds_d1Vs ds_d1Vt) = 
    (if (is$Example.A ds_d1Vs) 
     then Example.E 
     else (if (is$Example.C ds_d1Vt) 
           then Example.F 
           else Example.E))))
```

**Comment by ranjitjhala** [comment]
Hmm, interesting. 

This is the GHC core in the two cases, **after** Transform-Rec CoreBinds (before that step they are basically identical)  

```haskell
 Example.silly :: Example.Foo -> Example.Alias -> Example.Baz
 [LclIdX]
 Example.silly
   = \ (ds_d1Vy :: Example.Foo) (ds_d1Vz :: Example.Alias) ->
       case ds_d1Vy of {
         Example.A -> break<1>() Example.E;
         Example.B ->
           case ds_d1Vz of {
             __DEFAULT -> break<3>() Example.E;
             Example.C -> break<2>() Example.F
           }
       }
```

versus (perhaps due to the case-expand)

```haskell
 Example.silly :: Example.Foo -> Example.Bar -> Example.Baz
 [LclIdX]
 Example.silly
   = \ (ds_d1Vs :: Example.Foo) (ds_d1Vt :: Example.Bar) ->
       case ds_d1Vs of lq_anf$##7205759403792801042 {
         Example.A -> break<1>() Example.E;
         Example.B ->
           case ds_d1Vt of lq_anf$##7205759403792801043 {
             Example.C -> break<2>() Example.F;
             Example.D -> break<3>() Example.E
           }
       }
```

My guess is that the **case-expand** code (`ANF.expandDefaultCase'`) 
is failing to call *expandTypeAliases* before looking up the extra cases.

**Comment by ranjitjhala** [comment]
@nikivazou can you confirm that this in fact also handles #1225 ?

**Comment by nikivazou** [comment]
Thanks for the fix Ranjit!

@wkunkel can you check if this fixes #1225? (Kind of crazy schedule these days...) The code for #1225 is a simplification of our derivations repo.

**Comment by rosekunkel** [comment]
I confirmed that this fixes #1225
